### ISSUE TITLE
Expose ways to signal end of stream and failure?

### ISSUE THREAD

**Issue by xkxx** [mostjs/adapter#1]
Hi there,

Are adapters only intended for infinite streams, or are we meaning to support signaling end of streams and failure in the future?
I can think of a few ways to support the usage (for example, changing the first param to be a tuple(induce, end, fail)), but that would be a breaking change on the API, and so it would be ideal to consider the use case early on if possible.

Thanks,

xkxx

**Comment by briancavalier** [comment]
Hey @xkxx, thanks for asking.  We wanted to start as simple as possible, since imperative adapters like this have some inherent risks (like producer/consumer race conditions), and then let concrete use cases due any future changes.

Are there specific use cases where you need to imperatively signal end and/or failure?

**Comment by xkxx** [comment]
A common use case that I have in mind at the moment is converting node.js Streams. It's conceivable that at the end of the stream the client wants to run some kind of side effect at the end of stream (cleaning up, for example), or handle errors. It's also useful for converting promises, although I'm not sure if that's something that @most/core already addresses somewhere.

Adapters seem very similar with [create](https://github.com/mostjs/create), but with a different API, and I think create doesn't work with @most/core at the moment.

**Comment by briancavalier** [comment]
Thanks for that info, @xkxx.  This got a bit long, but I hope it's helpful.

#### tl;dr

Consider just implementing the [Stream interface](https://mostcore.readthedocs.io/en/latest/api.html#stream), which is a simple way to solve many use cases and allows controlling end and failure.  If that won't work for your use case, let's discuss whether adding end/fail capabilities to `@most/adapter` is what we should do.

---

#### More detail

One of the things we think makes mostjs interesting is its architectural simplicity.  One example is the [`Stream` interface](https://mostcore.readthedocs.io/en/latest/api.html#stream).  It's possible to create a huge variety of interesting things by implementing an interface with that single method.  All of mostjs is essentially built on that one method signature.

Quite often, the simplest way to make a new kind of event stream, such as one that can wrap a node.js data stream, is just to write an implementation of the Stream interface to do what you need.  This approach has some nice advantages, and avoids many of the pitfalls inherent in an imperative API like `@most/adapter`.

For example, here's a mostjs-community project that does just that: https://github.com/mostjs-community/most-node-streams.  It hasn't been updated to be compatible with `@most/core`, but the idea is the same, and it'd be trivial to update.  The existing [`fromPromise`](https://mostcore.readthedocs.io/en/latest/api.html#frompromise) is also built simply from the `Stream` interface.

One might wonder, then, why `@most/adapter` exists 😄.  There are cases where wrapping a stream around certain structures is hard, clumsy, or simply infeasible.  One example is React class-based components.  It's much easier to use something like `@most/adapter` than to wrap a Stream entirely around a React class component and still have it be usable within a larger React app.

As a starting, I'd recommend taking a look at the [Stream](https://mostcore.readthedocs.io/en/latest/api.html#stream), [Sink](https://mostcore.readthedocs.io/en/latest/api.html#sink), and [Disposable](https://mostcore.readthedocs.io/en/latest/api.html#disposable) types.  See if implementing Stream will work for your use cases.

Another nice example of a simple implementation of the Stream interface is [`@most/dom-event`](https://github.com/mostjs/dom-event/blob/master/src/index.js#L59).  Additionally, checkout other [mostjs-community projects](https://github.com/mostjs-community) for examples.

I hope that helps!

**Comment by briancavalier** [comment]
@xkxx Please let us know if the above is helpful and if implementing the Stream interface is a good solution for your use case.

If we haven't heard from you in a few days, I'll close this, but feel free to reopen if you need.

**Comment by xkxx** [comment]
Hi @briancavalier,

Thanks for the very extensive and thoughtful reply! I thought about replying that providing a complete imperative API would much simplify the work of converting existing data structures to streams. But I considered what you said more, and I agree that for most cases I can think of (promise, node.js stream, etc) error & termination require clean up logic that is specific to the type of source and is better done through implementing the complete stream interface. Coming from a rxjs background where the default and easy way is to use a subject, implementing stream directly seems like more upfront work, but I do agree that it is the right thing to do.

I'm also surprised how easy it is to implement a new stream, and if we want to encourage that as the primary way of creating new streams, it would be very helpful to have a step-by-step guide in the docs of how to do it (and do it properly). It would also be helpful to preface `adapter` with info on when and when not to use it and point to the aforementioned guide as the preferred alternative. But that probably belongs to a different issue now.

**Comment by briancavalier** [comment]
@xkxx I'm glad that was helpful.  Thanks for your feedback about documentation.

> it would be very helpful to have a step-by-step guide in the docs 

Agreed!

> It would also be helpful to preface adapter with info on when and when not to use it

Yes, definitely.  In fact, @Frikki added this in #2, and we merged it just yesterday.  There's a ["when you should use it"](https://github.com/mostjs/adapter#when-you-should-use-it) section in the README now.

Cheers
